### PR TITLE
hubble: Switch GetEventsChannel to new consumer interface

### PR DESCRIPTION
**Pull request by gandro**
This is the third of three PRs to remove the need to go through GOB encoding when accessing agent notifications in Hubble. The three PRs depend on each other, but I split them for easier reviewability.

1. monitor: Add new consumer interface which avoids gob encoded messages #12710
2. hubble: Add new monitor consumer implementation #12711
3. hubble: Switch GetEventsChannel to new consumer interface (**This PR**)

---

This commit switches the observer to use the new consumer interface
instead of relying on the monitor listener to feed the events queue.

This removes the need for GOB decoding in the parsers and decouples
the L3/L4 and L7 parsers from the old monitor payload type.

In addition, this adds support for keeping track of events dropped
in the Hubble events queue by the monitor consumer. The old monitor
listener code is removed.

### CODE REVIEW

**Comment by gandro** [comment]
test-me-please

File: pkg/hubble/observer/local_observer.go [hunk 1]
```diff
@@ -143,21 +142,21 @@ func (s *LocalObserverServer) Start() {
...
 				continue nextEvent
 			}
 		}
 
-		ev, err := s.payloadParser.Decode(pl)
+		ev, err := s.payloadParser.Decode(monitorEvent)
 		if err != nil {
-			if !parserErrors.IsErrInvalidType(err) {
-				s.log.WithError(err).WithField("data", pl.Data).Debug("failed to decode payload")
+			if err != parserErrors.ErrUnknownEventType || true {
```
Comment (rolinh) [inline pkg/hubble/observer/local_observer.go]:
```suggestion
			if !errors.Is(err, parserErrors.ErrUnknownEventType) {
```
I assume that `|| true` is a leftover from local testing.

File: pkg/hubble/parser/parser_test.go [hunk 2]
```diff
@@ -17,18 +17,17 @@
...
-
 	flowpb "github.com/cilium/cilium/api/v1/flow"
 	v1 "github.com/cilium/cilium/pkg/hubble/api/v1"
+	observerTypes "github.com/cilium/cilium/pkg/hubble/observer/types"
 	"github.com/cilium/cilium/pkg/hubble/parser/errors"
 	"github.com/cilium/cilium/pkg/hubble/testutils"
 	"github.com/cilium/cilium/pkg/monitor"
 	"github.com/cilium/cilium/pkg/monitor/api"
 	"github.com/cilium/cilium/pkg/proxy/accesslog"
+	"io/ioutil"
```
Comment (rolinh) [inline pkg/hubble/parser/parser_test.go]: nit: you probably didn't mean to group stdlib imports with the others.

**Review by rolinh — CHANGES_REQUESTED** [review]
Looks good overall; just a change for something you probably didn't mean to commit ;)

File: pkg/hubble/observer/local_observer.go (hunk 1 above)
Comment (gandro) [reply pkg/hubble/observer/local_observer.go]: :scream: true!

File: pkg/hubble/parser/parser_test.go (hunk 2 above)
Comment (gandro) [reply pkg/hubble/parser/parser_test.go]: Yes.

File: pkg/hubble/observer/local_observer.go (hunk 1 above)
Comment (gandro) [reply pkg/hubble/observer/local_observer.go]: Fixed

**Comment by gandro** [comment]
test-me-please

**Review by michi-covalent — APPROVED** [review]
💯

**Comment by gandro** [comment]
CI failures look legitimate. There seems to be an issue with L7 visibility with this PR.

**Comment by gandro** [comment]
test-me-please

File: pkg/hubble/observer/local_observer.go [hunk 3]
```diff
@@ -44,26 +45,6 @@ import (
...
-	///GetRingBuffer returns the underlying ring buffer to parsed events.
-	GetRingBuffer() *container.Ring
-	// GetStopped returns a channel that gets closed at the end of the
-	// main server loop after all the events have been processes. Used
-	// in unit testing.
-	GetStopped() chan struct{}
-	// GetLogger returns the logger assigned to this gRPC server.
-	GetLogger() logrus.FieldLogger
-}
-
```
Comment (rolinh) [inline pkg/hubble/observer/local_observer.go:66]: I like this diff :)
However, several methods comments of `LocalObserverServer` [1] now have incorrect comments such as "Start implements GRPCServer.Start".

[1]: This struct should really be renamed at this point but this is outside the scope of this PR.

**Review by rolinh — APPROVED** [review]
No more GOB decoding; that's awesome :tada: 

LGTM overall, just added a nit for now outdated method comments.

File: pkg/hubble/observer/local_observer.go (hunk 3 above)
Comment (gandro) [reply pkg/hubble/observer/local_observer.go:66]: Yeah let's update the comments and naming in a separate PR.
